drivers/mipi_dbi/mipi_dbi_spi.c

```diff
@@ -94,34 +98,103 @@ static int mipi_dbi_spi_write_helper(const struct device *dev,
 			}
 		}
 	} else if (dbi_config->mode == MIPI_DBI_MODE_SPI_4WIRE) {
-		/* 4 wire mode is much simpler. We just toggle the
-		 * command/data GPIO to indicate if we are sending
-		 * a command or data
-		 */
-		buffer.buf = &cmd;
-		buffer.len = sizeof(cmd);
-
-		if (cmd_present) {
-			/* Set CD pin low for command */
-			gpio_pin_set_dt(&config->cmd_data, 0);
-			ret = spi_write(config->spi_dev, &dbi_config->config,
-					&buf_set);
-			if (ret < 0) {
-				goto out;
+		if (config->xfr_min_bits == MIPI_DBI_SPI_XFR_8BIT) {
+			/* 4 wire mode is much simpler. We just toggle the
+			 * command/data GPIO to indicate if we are sending
+			 * a command or data (the default).
+			 */
+			buffer.buf = &cmd;
+			buffer.len = sizeof(cmd);
+
+			if (cmd_present) {
+				/* Set CD pin low for command */
+				gpio_pin_set_dt(&config->cmd_data, 0);
+				ret = spi_write(config->spi_dev, &dbi_config->config,
+						&buf_set);
+				if (ret < 0) {
+					goto out;
+				}
 			}
-		}
 
-		if (len > 0) {
-			buffer.buf = (void *)data_buf;
-			buffer.len = len;
-
-			/* Set CD pin high for data */
-			gpio_pin_set_dt(&config->cmd_data, 1);
-			ret = spi_write(config->spi_dev, &dbi_config->config,
-					&buf_set);
-			if (ret < 0) {
-				goto out;
+			if (len > 0) {
+				buffer.buf = (void *)data_buf;
+				buffer.len = len;
+
+				/* Set CD pin high for data */
+				gpio_pin_set_dt(&config->cmd_data, 1);
+				ret = spi_write(config->spi_dev, &dbi_config->config,
+						&buf_set);
+				if (ret < 0) {
+					goto out;
+				}
+			}
+		} else if (config->xfr_min_bits == MIPI_DBI_SPI_XFR_16BIT) {
+			/* 4 wire mode with toggle the command/data GPIO
+			 * to indicate if we are sending a command or data
+			 * but send 16-bit blocks (bit stuffing).
+			 */
+			uint16_t data16;
+
+			if (cmd_present) {
+				data16 = sys_cpu_to_be16(cmd);
+				buffer.buf = &data16;
+				buffer.len = sizeof(data16);
+
+				/* Set CD pin low for command */
+				gpio_pin_set_dt(&config->cmd_data, 0);
+				ret = spi_write(config->spi_dev, &dbi_config->config,
+						&buf_set);
+				if (ret < 0) {
+					goto out;
+				}
+
+				/* iterate command data */
+				for (int i = 0; i < len; i++) {
+					data16 = sys_cpu_to_be16(data_buf[i]);
+
+					/* Set CD pin high for data */
+					gpio_pin_set_dt(&config->cmd_data, 1);
```

Comment:
You likely only need to set the CD pin logic high here once, not in each loop iteration

Reply:
done